### PR TITLE
streamaggr: skip samples that are older than total value startTime

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**

### CODE REVIEW

File: lib/streamaggr/streamaggr.go [hunk 1]
```diff
@@ -775,11 +799,15 @@ func (a *aggregator) Push(tss []prompbmarshal.TimeSeries, matchIdxs []byte) {
...
-			})
+			if a.discardSamplesOlderThanSecs > 0 && sample.Timestamp < discardLimitMsec {
+				a.oldDiscardedSamples.Inc()
+			} else {
+				samples = append(samples, pushSample{
+					key:       key,
+					value:     sample.Value,
+					timestamp: sample.Timestamp,
+				})
+			}
```
Comment (hagen1778) [inline lib/streamaggr/streamaggr.go:810]:
```suggestion
			if a.discardSamplesOlderThanSecs > 0 && sample.Timestamp < discardLimitMsec {
				a.oldDiscardedSamples.Inc()
				continue
			}
			samples = append(samples, pushSample{
				key:       key,
				value:     sample.Value,
				timestamp: sample.Timestamp,
			})
	
```

File: lib/streamaggr/streamaggr.go [hunk 2]
```diff
@@ -192,6 +193,9 @@ type Config struct {
 	// OutputRelabelConfigs is an optional relabeling rules, which are applied
 	// on the aggregated output before being sent to remote storage.
 	OutputRelabelConfigs []promrelabel.RelabelConfig `yaml:"output_relabel_configs,omitempty"`
+
+	// DiscardSamplesOlderThan interval older which series are discarded
+	DiscardSamplesOlderThan string `yaml:"discard_samples_older_than"`
```
Comment (hagen1778) [inline lib/streamaggr/streamaggr.go:198]: To what value would you recommend setting this parameter?

File: lib/streamaggr/streamaggr_test.go [hunk 3]
```diff
@@ -983,6 +984,9 @@ func mustParsePromMetrics(s string) []prompbmarshal.TimeSeries {
 				Value: tag.Value,
 			})
 		}
+		if row.Timestamp == 0 {
```
Comment (hagen1778) [inline lib/streamaggr/streamaggr_test.go:987]: can we add tests to check discard functionality specifically?

File: lib/streamaggr/total.go [hunk 4]
```diff
@@ -23,13 +23,7 @@ type totalAggrState struct {
 	// Time series state is dropped if no new samples are received during stalenessSecs.
 	//
 	// Aslo, the first sample per each new series is ignored during stalenessSecs even if keepFirstSample is set.
-	// see ignoreFirstSampleDeadline for more details.
 	stalenessSecs uint64
-
-	// The first sample per each new series is ignored until this unix timestamp deadline in seconds even if keepFirstSample is set.
-	// This allows avoiding an initial spike of the output values at startup when new time series
-	// cannot be distinguished from already existing series. This is tracked with ignoreFirstSampleDeadline.
-	ignoreFirstSampleDeadline uint64
```
Comment (hagen1778) [inline lib/streamaggr/total.go:32]: I think this functionality isn't related to functionality of discarding "too old" samples. It is applied only to the first recorded sample. While `discard_samples_older_than` is applied to the whole data stream. Can we return it back?

**Review by hagen1778 — CHANGES_REQUESTED** [review]

File: lib/streamaggr/dedup.go [hunk 5]
```diff
@@ -209,8 +209,9 @@ func (das *dedupAggrShard) flush(ctx *dedupFlushCtx, f func(samples []pushSample
 	dstSamples := ctx.samples
 	for key, s := range m {
 		dstSamples = append(dstSamples, pushSample{
-			key:   key,
-			value: s.value,
+			key:       key,
+			value:     s.value,
+			timestamp: s.timestamp,
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply lib/streamaggr/dedup.go:214]: it's rebased

[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply lib/streamaggr/dedup.go:214]: it was not necessary before as setting this timestamp here is needed for processing in later aggregators

**Pull request closed without merging**